src/main/java/com/limechain/babe/Authorship.java

```diff
@@ -10,54 +15,194 @@
+
+                return buildSecondaryVrfPreDigest(
+                        randomness,
+                        slotNumber,
+                        epochIndex,
+                        keyPair,
+                        authorityIndex
+                );
+            } else {
+                log.log(Level.FINE, "Secondary Plain slot successfully claimed for slot number: {}", slotNumber);
+
+                return new BabePreDigest(
+                        PreDigestType.BABE_SECONDARY_PLAIN,
+                        authorityIndex.longValue(),
+                        slotNumber,
+                        null,
+                        null
+                );
+            }
+        }
+
+        return null;
+    }
+
+    private static BabePreDigest buildSecondaryVrfPreDigest(final byte[] randomness,
+                                                            final BigInteger slotNumber,
+                                                            final BigInteger epochIndex,
+                                                            final Schnorrkel.KeyPair keyPair,
+                                                            final Integer authorityIndex) {
+
+        var transcript = makeTranscript(randomness, slotNumber, epochIndex);
+        VrfOutputAndProof vrfOutputAndProof = Schnorrkel.getInstance().vrfSign(keyPair, transcript);
+
+        return new BabePreDigest(
+                PreDigestType.BABE_SECONDARY_VRF,
+                authorityIndex.longValue(),
+                slotNumber,
+                vrfOutputAndProof.getOutput(),
+                vrfOutputAndProof.getProof()
+        );
+    }
+
+    private static Integer getSecondarySlotAuthor(final byte[] randomness,
+                                                  final BigInteger slotNumber,
+                                                  final List<Authority> authorities) {
+        if (authorities.isEmpty()) return null;
+
+        byte[] concat = ByteArrayUtils.concatenate(randomness, slotNumber.toByteArray());
+
+        Blake2b.Blake2b256 blake2b256 = new Blake2b.Blake2b256();
+        byte[] hash = blake2b256.digest(concat);
+
+        var rand = LittleEndianUtils.fromLittleEndianByteArray(hash);
+        var authoritiesCount = BigInteger.valueOf(authorities.size());
+        var authorityIndex = rand.mod(authoritiesCount);
+
+        if (authorityIndex.compareTo(authoritiesCount) < 0) {
+            return authorityIndex.intValue();
         }
 
         return null;
     }
 
     // threshold = 2^128 * (1 - (1 - c) ^ (authority_weight / sum(authorities_weights)))
-    public static BigInteger calculatePrimaryThreshold(
+    private static BigInteger calculatePrimaryThreshold(
             @NotNull final Pair<BigInteger, BigInteger> constant,
             @NotNull final List<Authority> authorities,
-            final int authorityIndex) {
+            final Integer authorityIndex) {
```

Comment:
Is this nullable now?

Reply:
My idea was to use Integer in order to use Integer.longValue() when adding the index to the PreDigest, but I agree is better to stay with the primitive int type.